aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -0,0 +1,149 @@
+
+        let original_event_structs =
+            extract_event_metadata_from_module(session, &module.self_id())?;
+
+        for member in original_event_structs {
+            // We don't need to re-validate new_members above.
+            if !new_event_structs.remove(&member) {
+                metadata_validation_err("Invalid change in event attributes")?;
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Validate all the `0x1::event::emit` calls have the struct defined in the same module with event
+/// attribute.
+pub(crate) fn validate_emit_calls(
+    event_structs: &HashSet<String>,
+    module: &CompiledModule,
+) -> VMResult<()> {
+    for fun in module.function_defs() {
+        if let Some(code_unit) = &fun.code {
+            for bc in &code_unit.code {
+                if let Bytecode::CallGeneric(index) = bc {
+                    let func_instantiation = &module.function_instantiation_at(*index);
+                    let func_handle = module.function_handle_at(func_instantiation.handle);
+                    let module_handle = module.module_handle_at(func_handle.module);
+                    let module_addr = module.address_identifier_at(module_handle.address);
+                    let module_name = module.identifier_at(module_handle.name);
+                    let func_name = module.identifier_at(func_handle.name);
+                    if module_addr
+                        != &AccountAddress::from_hex_literal("0x1").expect("valid address")
+                        || module_name.as_str() != EVENT_MODULE_NAME
+                        || func_name.as_str() != EVENT_EMIT_FUNCTION_NAME
+                    {
+                        continue;
+                    }
+                    let param = module
+                        .signature_at(func_instantiation.type_parameters)
+                        .0
+                        .first()
+                        .ok_or_else(|| {
+                            metadata_validation_error(
+                                "Missing parameter for 0x1::event::emit function",
+                            )
+                        })?;
+                    match param {
+                        StructInstantiation(index, _) | Struct(index) => {
+                            let struct_handle = &module.struct_handle_at(*index);
+                            let struct_name = module.identifier_at(struct_handle.name);
+                            if struct_handle.module != module.self_handle_idx() {
+                                metadata_validation_err(format!("{} passed to 0x1::event::emit function is not defined in the same module", struct_name).as_str())
+                            } else if !event_structs.contains(struct_name.as_str()) {
+                                metadata_validation_err(format!("#[event] attribute is missing for {} passed to 0x1::event::emit function", struct_name).as_str())
+                            } else {
```

Comment:
`Missing #[event] attribute on {}. The #[event] attribute is required for all structs passed into 0x1::event::emit`.